### PR TITLE
PaliGemma fix attention mask for finetunes

### PR DESCRIPTION
**Pull request by probicheaux**
# What does this PR do?

Previously, the PaliGemma attention mask was a 4d attention mask of all 1s, which preempted any downstream causal mask creation by Gemma. [PaliGemma has full attention over the image tokens and the prefix of the prompt, and causal attention over the suffix.](https://github.com/google-research/big_vision/blob/main/big_vision/configs/proj/paligemma/README.md) This PR implements that logic, allowing for finetuning PaliGemma without label leakage.

(Additionally, I noticed that using the GemmaTokenizerFast led to incorrect tokenization of tokens like <loc0000> character by character, whereas using GemmaTokenizer did proper tokenization. Be sure to set `use_fast=False` when instantiating the PaliGemma processor. I'll make an issue for this observation
edit: ah, [looks like that is a known issue?](https://github.com/huggingface/transformers/issues/30824)).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @molbap @merveenoyan

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Can you provide a reproducer for:
1. the tokenizer issues: it's probably just about the added tokens and the conversion 
2. the causal mask after the prefill is equivalent to an all 1s mask because you only have one new token and the PKV has the "strored" information on the previous attention pattern so I am sure I understand what the issue is here?

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 1]
```diff
@@ -282,6 +283,47 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
...
+        prefix_and_image_and_pad_mask = range_tensor < indices.view(-1, 1, 1)
+        prefix_and_image_and_pad_mask = prefix_and_image_and_pad_mask.expand(-1, mask.size(1), -1)
+        causal_mask = causal_mask[None, :].expand(input_ids.shape[0], -1, -1)
+        causal_mask[prefix_and_image_and_pad_mask] = 0
+        causal_mask = causal_mask[:, None, :, :].expand(-1, 1, -1, -1)
+        if attention_mask is not None:
+            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+            mask_length = attention_mask.shape[-1]
+            padding_mask = attention_mask[:, None, None, :]
+            padding_mask = padding_mask == 0
```
Comment (molbap) [inline src/transformers/models/paligemma/modeling_paligemma.py:316]:
```suggestion
            padding_mask = attention_mask[:, None, None, :].eq(0)
```

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 2]
```diff
@@ -282,6 +283,47 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
...
+        causal_mask = torch.full(
+            (sequence_length, target_length),
+            fill_value=min_dtype,
+            dtype=dtype,
+            device=device,
+        )
+        if sequence_length != 1:
+            causal_mask = torch.triu(causal_mask, diagonal=1)
+
+        mask = input_ids == self.config.prefix_suffix_separator_index
```
Comment (molbap) [inline src/transformers/models/paligemma/modeling_paligemma.py:301]: This is not very robust unfortunately: one or several `\n` tokens might be present in the prefix. Could this depend on the labels passed instead?

**Review by molbap — COMMENTED** [review]
Good catch, main comment is that separation between prefix and suffix is brittle as-is - we want to merge this fix as soon as possible though!

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 3]
```diff
@@ -282,6 +283,47 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
...
+        # Get the index of the first \n in each row
+        indices = mask.int().cumsum(dim=1).eq(1).int().argmax(dim=1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (probicheaux) [reply src/transformers/models/paligemma/modeling_paligemma.py:303]: The cumsum catches the first \n and splits off that. But if we want to do something with the labels instead we can do that change. 

Actually, maybe the cumsum isn't necessary you are right

**Comment by probicheaux** [comment]
I accidentally ran style on the whole repo and now the tests are failing cause I had to force push away updating like every file (இ﹏இ`｡)

I think all the style errors are in code not touched by this change

**Comment by probicheaux** [comment]
Anyways OK I updated how we check where to split using a new index in the labels. Constructing the right labels array is now getting to be sort of tricky for new users, so now the PaliGemmaProcessor returns labels in BatchEncoding for easier finetuning. This is sort of necessary because the prefix and suffix might change length during tokenization and it'd be annoying to try to find the right index of that \n. Hope y'all are ok with this idea!!!

**Review by ArthurZucker — COMMENTED** [review]
A lot of the logic you are re-implementing is natively supported in the tokenizer with the `token_type_ids`. In this case I think it would be wise to leverage `create_token_type_ids_from_sequences` checkout codegen for example.

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 4]
```diff
@@ -99,12 +110,15 @@ def __init__(self, image_processor=None, tokenizer=None):
...
         self.image_token_id = tokenizer.convert_tokens_to_ids(IMAGE_TOKEN)
+        self.ignore_index = ignore_index
+        self.prefix_suffix_separator_index = prefix_suffix_separator_index
 
         super().__init__(image_processor, tokenizer)
 
     def __call__(
         self,
         text: Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]] = None,
+        suffix: Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]] = None,
```
Comment (molbap) [inline src/transformers/models/paligemma/processing_paligemma.py:121]:
```suggestion
        labels: Optional[Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]]] = None,
```

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 5]
```diff
@@ -135,6 +151,9 @@ def __call__(
                 The sequence or batch of sequences to be encoded. Each sequence can be a string or a list of strings
                 (pretokenized string). If the sequences are provided as list of strings (pretokenized), you must set
                 `is_split_into_words=True` (to lift the ambiguity with a batch of sequences).
+            suffix (`str`, `List[str]`, `List[List[str]]`):
+                The suffix or batch of suffixes to be encoded. Only necessary for finetuning. See https://github.com/google-research/big_vision/blob/main/big_vision/configs/proj/paligemma/README.md
+                for more information.
```
Comment (molbap) [inline src/transformers/models/paligemma/processing_paligemma.py:156]:
```suggestion
            labels (`str`, `List[str]`, `List[List[str]]`):
                The labels suffixes or batch of suffixes to be encoded. Only necessary for finetuning. See https://github.com/google-research/big_vision/blob/main/big_vision/configs/proj/paligemma/README.md
                for more information.
```

Comment (molbap) [reply src/transformers/models/paligemma/processing_paligemma.py:156]: Overall replacing suffix with labels where it's use, I think

File: src/transformers/models/paligemma/processing_paligemma.py [hunk 6]
```diff
@@ -223,26 +246,109 @@ def __call__(
...
 
-            text_inputs = self.tokenizer.pad(
-                {"input_ids": concatenated_ids, "attention_mask": concatenated_attention_masks},
-                max_length=max_length,
-                padding=padding,
-                return_tensors=return_tensors,
-            )
         else:
-            text_inputs = self.tokenizer(
+            inputs = self.tokenizer(
```
Comment (molbap) [inline src/transformers/models/paligemma/processing_paligemma.py:253]: Here I think the key is using the `text_pair` option in the tokenizer along with `return_token_type_ids=True`. 
So
```python
text = ["Is the witholded tax rate equal to 31%?"]
labels = ["Yes"]
 self.tokenizer(text=text, text_pair=labels, return_token_type_ids=True)
```
This will give the additional key to `inputs` of the form `'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0, 1, 1]]`  and will have also correct shape for batched outputs. 

Then you already have your mask from `token_type_ids` and don't need to specify the `training` bool flag below. The token for newline character has to be inserted before the first 1, and assigned the token type id 0 so that it's part of the prompt with full block attention.

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 7]
```diff
@@ -306,23 +347,24 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
...
-        )
-
         # position_ids = torch.arange(0, sequence_length, device=input_ids.device).expand(batch_size, -1)
         # position_ids = torch.where(input_ids == self.pad_token_id, torch.ones_like(position_ids), position_ids)
         position_ids = (attention_mask.cumsum(-1)).masked_fill_((attention_mask == 0), 1)
 
         if labels is not None:
+            causal_mask = self.construct_causal_mask_with_block_attention(
+                attention_mask, labels, text_mask, inputs_embeds
+            )
```
Comment (molbap) [inline src/transformers/models/paligemma/modeling_paligemma.py:357]: Here, the `token_type_ids` can be passed from the forward since `inputs` are unpacked and will contain them. Then, you already have your mask to build the causal mask!

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 8]
```diff
@@ -282,6 +283,46 @@ def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_m
...
+            fill_value=min_dtype,
+            dtype=dtype,
+            device=device,
+        )
+        if sequence_length != 1:
+            causal_mask = torch.triu(causal_mask, diagonal=1)
+
+        mask = labels == self.config.prefix_suffix_separator_index
+        # Get the index of the first \n in each row
+        indices = mask.int().argmax(dim=1, keepdim=True)
```
Comment (molbap) [inline src/transformers/models/paligemma/modeling_paligemma.py:303]: Assuming you're getting the `token_type_ids` here, you just have to do something like that
```suggestion
        indices = (token_type_ids == 1).int().argmax(dim=1)
```

**Review by molbap — COMMENTED** [review]
Hey @probicheaux , left another review! I think your causal mask building is fine, we just need to remove `training` flag from the processor and pass the labels explicitly. Let me know if you have time to add the suggestions!

**Comment by probicheaux** [comment]
Merged as part of https://github.com/huggingface/transformers/pull/30967
